lib/que/job.rb

```diff
@@ -1,183 +1,79 @@
+      # we'll keep it for compatibility.
+      inserted_job =
+        Que.execute(:insert_job, [*attrs.values_at(*JOB_OPTIONS), args]).first
 
-    def destroy
-      Que.execute :destroy_job, attrs.values_at(:queue, :priority, :run_at, :job_id)
-      @destroyed = true
+      job = new(inserted_job)
+      # TODO: _run -> run_and_destroy(*inserted_job[:args])
+      job._run if Que.mode == :sync
+      job
     end
 
-    def self.handle_job_failure(error, job)
-      count    = job[:error_count].to_i + 1
-
-      # Prefactor: changes mean we have to fetch default from self || superclass
-      # interval = klass && klass.respond_to?(:retry_interval) && klass.retry_interval || retry_interval
-      interval = retry_interval || superclass.respond_to?(:retry_interval) && superclass.retry_interval
-      delay    = interval.respond_to?(:call) ? interval.call(count) : interval
-      message  = "#{error.message}\n#{error.backtrace.join("\n")}"
-      Que.execute :set_error, [count, delay, message] + job.values_at(:queue, :priority, :run_at, :job_id)
+    def self.run(*args)
+      new(args: args).tap { |job| job.run(*args) }
     end
 
-    class << self
-      attr_reader :retry_interval
-
-      def enqueue(*args)
-        if args.last.is_a?(Hash)
-          options   = args.pop
-          queue     = options.delete(:queue) || '' if options.key?(:queue)
-          job_class = options.delete(:job_class)
-          run_at    = options.delete(:run_at)
-          priority  = options.delete(:priority)
-          retryable = options.delete(:retryable)
-          args << options if options.any?
-        end
-
-        attrs = {:job_class => job_class || to_s, :args => args}
-
-        warn "@default_run_at in #{to_s} has been deprecated and will be removed in Que version 1.0.0. Please use @run_at instead." if @default_run_at
-
-        if t = run_at || @run_at && @run_at.call || @default_run_at && @default_run_at.call
-          attrs[:run_at] = t
-        end
-
-        warn "@default_priority in #{to_s} has been deprecated and will be removed in Que version 1.0.0. Please use @priority instead." if @default_priority
-
-        if p = priority || @priority || @default_priority
-          attrs[:priority] = p
-        end
-
-        if q = queue || @queue
-          attrs[:queue] = q
-        end
-
-        if retryable.nil?
-          attrs[:retryable] = true
-        else
-          attrs[:retryable] = retryable
-        end
-
-        if Que.mode == :sync && !t
-          # We need to run this through JSON serializers to ensure we get a pure JSON
-          # result. Que then does some seriously weird stuff with the `json_converter`
-          # to make the hashes we pass to Que jobs respond to both symbols and strings.
-          # We should use the json_converter here to be consistent.
-          run(*Que.json_converter.call(parse_json(generate_json(attrs)))[:args])
```

Comment:
We used to have to do this horrible, horrible thing here. In refactoring have we re-introduced the issue mentioned in the comment?

Reply:
We have not! We no longer need to do this because we're much more consistent with how we handle jobs. Instead of enqueue just running the job, it now adds it to the database and we run it just as if we'd picked it up in a worker. This simplifies how we handle jobs and ensures we don't have inconsistent serialization.

Reply:
This is a much better approach, super happy we dont rely on that line of code anymore 😹